src/System.Data.SqlClient/tests/Tools/TDS/TDS.Clients/GenericTDSClient.cs

```diff
@@ -0,0 +1,594 @@
+            {
+                // Doing SQL authentication
+                loginToken = CompleteSqlAuthRequest(preLoginResponse);
+            }
+            else if (Arguments.SendCompleteLogin)
+            {
+                // Doing complete federated authentication.				
+                loginToken = CompleteFedAuthLoginRequest(preLoginResponse);
+            }
+            else
+            {
+                // Doing fedauth request without fedauth token in it.
+                loginToken = CompleteFedAuthInfoRequest(preLoginResponse);
+            }
+
+            // Log request
+            TDSUtilities.Log(Arguments.Log, "Request", loginToken);
+
+            // Serialize login token into the request packet
+            requestMessage.Add(loginToken);
+
+            return requestMessage;
+        }
+
+
+        /// <summary>
+        /// Federated authentication info received.
+        /// </summary>
+        /// <param name="response">TDS response message from server.</param>
+        /// <returns>TDS message to send to the server.</returns>
+        public virtual TDSMessage OnFederatedAuthenticationInfoResponse(TDSMessage response)
+        {
+            // Create new request message.
+            TDSMessage requestMessage = new TDSMessage(TDSMessageType.FederatedAuthentication);
+
+            // Take the fedauth info from the response message.
+            TDSFedAuthInfoToken fedAuthResponse = response[0] as TDSFedAuthInfoToken;
+
+            // Log the response message from the server.
+            TDSUtilities.Log(Arguments.Log, "Response", fedAuthResponse);
+
+            // Create fedauth message token.
+            TDSFedAuthMessageToken fedAuthMessageToken = new TDSFedAuthMessageToken();
+
+            // Get the nonce from the context that is received by prelogin exchange.
+            fedAuthMessageToken.Nonce = (Context as GenericTDSClientContext).ServerNonce;
+
+            // Generatin fed auth Data.
+            // Note: In reality IDCRL calls should be done in order to get this token.
+            fedAuthMessageToken.Data = _GenerateRandomBytes(32);
+
+            // Log the request message.
+            TDSUtilities.Log(Arguments.Log, "Request", fedAuthMessageToken);
+
+            // Serialize fed auth message token into request packet.
+            requestMessage.Add(fedAuthMessageToken);
+
+            // Change the client state to complete login sent
+            State = TDSClientState.CompleteLogin7Sent;
+
+            return requestMessage;
+        }
+
+        /// <summary>
+        /// Received login acknowledgement
+        /// </summary>
+        public virtual void OnCompleteLoginResponse(TDSMessage response)
+        {
+            // Prepare exception
+            Exception exception = null;
+
+            // Reset all messages
+            (Context as GenericTDSClientContext).Messages = new List<string>();
```

Comment:
>Context as GenericTDSClientContext [](start = 13, length = 34)

Nit: could store this into a local so it doesn't need to be cast on every access